File: progChallenges/september/permutations_generate.py
#!/usr/bin/env python3

def permutations(p,c):
    '''
    p: current permuation
    c: set of candidates
    '''
    #base case: we run out of candidates
    if not c:
        yield p
    else:
        for e in sorted(c):
            yield from permutations( 
                p + e,  #add element to permutation 
                c - {e} #remove element from candidates 
            )

def main():
    for p in permutations('',set('ABC')):
        print(''.join(p))

if __name__ == '__main__':
    main()

File: progChallenges/september/permutations_generate.cpp
// permutations_backtracking.cpp

#include <algorithm>
#include <iostream>
#include <set>
#include <string>

using namespace std;

// Functions

void permutations(string &p, set<char> &c) {
    // p: current permuation
    // c: set of candidates (sets in c++ are default ordered)
    if (c.empty()) {
        for (auto e : p) { cout << e;} cout << endl;
    } else {
        auto n = c;
        for (auto e : n) {
            p.push_back(e); //add element to permutation
            c.erase(e);     //remove element from candidates
            permutations(p,c);

            // backtrack
            c.insert(e);
            p.pop_back();
        }
    }
}

// Main execution

int main(int argc, char *argv[]) {
    string    p = "";
    set<char> c = {'A', 'B', 'C'};

    permutations(p, c);
    return 0;
}